Fix static data initialization deadlock on GIL

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34505 Fix static data initialization deadlock on GIL**

A thread could hold GIL when calling PythonRpcHandler::getInstance(),
meantime another thread could have been doing static data
initialization by calling `new PythonRpcHandler()`, inside of which GIL is
also required. Static data initialization is thread-safe, so the thread
holding the GIL will wait for the other thread to finish static data
initializating before going forward. Because the initialization can't
proceed without GIL, there is a deadlock. We ask the calling thread to
release GIL to avoid this situation.

Differential Revision: [D7490489](https://our.internmc.facebook.com/intern/diff/D7490489/)

xush6528 committed:
Fix static data initialization deadlock on GIL

A thread could hold GIL when calling PythonRpcHandler::getInstance(),
meantime another thread could have been doing static data
initialization by calling `new PythonRpcHandler()`, inside of which GIL is
also required. Static data initialization is thread-safe, so the thread
holding the GIL will wait for the other thread to finish static data
initializating before going forward. Because the initialization can't
proceed without GIL, there is a deadlock. We ask the calling thread to
release GIL to avoid this situation.

Differential Revision: [D7490489](https://our.internmc.facebook.com/intern/diff/D7490489/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix static data initialization deadlock on GIL"


A thread could hold GIL when calling PythonRpcHandler::getInstance(),
meantime another thread could have been doing static data
initialization by calling `new PythonRpcHandler()`, inside of which GIL is
also required. Static data initialization is thread-safe, so the thread
holding the GIL will wait for the other thread to finish static data
initializating before going forward. Because the initialization can't
proceed without GIL, there is a deadlock. We ask the calling thread to
release GIL to avoid this situation.

Differential Revision: [D7490489](https://our.internmc.facebook.com/intern/diff/D7490489/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix static data initialization deadlock on GIL"


A thread could hold GIL when calling PythonRpcHandler::getInstance(),
meantime another thread could have been doing static data
initialization by calling `new PythonRpcHandler()`, inside of which GIL is
also required. Static data initialization is thread-safe, so the thread
holding the GIL will wait for the other thread to finish static data
initializating before going forward. Because the initialization can't
proceed without GIL, there is a deadlock. We ask the calling thread to
release GIL to avoid this situation.

Differential Revision: [D7490489](https://our.internmc.facebook.com/intern/diff/D7490489/)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -409,31 +410,32 @@ If the future completes with an error, an exception is thrown.
-            kwargsDict.cast<py::kwargs>(),
-            c10::nullopt);
-        py::gil_scoped_release release;
+        c10::QualifiedName qualifiedName =
```
nit: auto& pythonRpcHandler = PythonRpcHandler::getInstance(), it has been used twice below

xush6528 replied:
Updated.

zhaojuanmao reviewed (approved):
one minor nit, please make sure tests pass before landing

xush6528 committed:
Update on "Fix static data initialization deadlock on GIL"


A thread could hold GIL when calling PythonRpcHandler::getInstance(),
meantime another thread could have been doing static data
initialization by calling `new PythonRpcHandler()`, inside of which GIL is
also required. Static data initialization is thread-safe, so the thread
holding the GIL will wait for the other thread to finish static data
initializating before going forward. Because the initialization can't
proceed without GIL, there is a deadlock. We ask the calling thread to
release GIL to avoid this situation.

Differential Revision: [D7490489](https://our.internmc.facebook.com/intern/diff/D7490489/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -342,9 +342,10 @@ If the future completes with an error, an exception is thrown.
+  module.def(
+      "_cleanup_python_rpc_handler",
+      []() { PythonRpcHandler::getInstance().cleanup(); },
+      py::call_guard<py::gil_scoped_release>());
```
Looks like the GIL will be immediately acquired again after entering `cleanup()`. Is there any specific reason for back-to-back release-acquire here?

mrshenli replied:
Oh I see, it is to avoid holding GIL for `PythonRpcHandler::getInstance().`

xush6528 replied:
both I and @zhaojuanmao had fallen into this pitfall, so I added intern assert to avoid people falling into again. But this change requires this callsite to release lock.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -409,31 +410,33 @@ If the future completes with an error, an exception is thrown.
+              kwargsDict.cast<py::kwargs>(),
+              c10::nullopt);
+        }
         DCHECK(!PyGILState_Check());
```
This is redundant?

xush6528 replied:
This is to revert back to https://github.com/pytorch/pytorch/pull/33610

This macro does not produce code unless in debug mode (gcc -g).

Not plan to change it.

xush6528 committed:
Update on "Fix static data initialization deadlock on GIL"


A thread could hold GIL when calling PythonRpcHandler::getInstance(),
meantime another thread could have been doing static data
initialization by calling `new PythonRpcHandler()`, inside of which GIL is
also required. Static data initialization is thread-safe, so the thread
holding the GIL will wait for the other thread to finish static data
initializating before going forward. Because the initialization can't
proceed without GIL, there is a deadlock. We ask the calling thread to
release GIL to avoid this situation.

Differential Revision: [D7490489](https://our.internmc.facebook.com/intern/diff/D7490489/)

[ghstack-poisoned]

xush6528 committed:
Update on "Fix static data initialization deadlock on GIL"


A thread could hold GIL when calling PythonRpcHandler::getInstance(),
meantime another thread could have been doing static data
initialization by calling `new PythonRpcHandler()`, inside of which GIL is
also required. Static data initialization is thread-safe, so the thread
holding the GIL will wait for the other thread to finish static data
initializating before going forward. Because the initialization can't
proceed without GIL, there is a deadlock. We ask the calling thread to
release GIL to avoid this situation.

Differential Revision: [D7490489](https://our.internmc.facebook.com/intern/diff/D7490489/)

[ghstack-poisoned]
